# Assignment-1 (Web Analytics)
**Name: Tej Dave** <br><br>
**Student ID: 8717454** <br><br>
**Due Date: 10-11-2021**

In [66]:
import pandas as pd
import numpy as np

# Step 1: Reading Dataset
* Creating columns list that are required. 
* Reading csv "Assignment1Data_Sample.csv".
* Using "Object ID" as an index.

In [67]:
#reading in dataset 

#col to use in list 
col_ls = ['Object ID','Department','Object Name','Title','Culture','Artist Nationality',
          'Object Begin Date','Object End Date','Medium','Credit Line','Country']

museum_df = pd.read_csv('Assignment1Data_Sample.csv',index_col='Object ID',usecols=col_ls)

#observing the head 
museum_df.head()

Department          Object Name               Title  \
Object ID                                                                      
1674       American Decorative Arts           Side Chair          Side Chair   
2430       American Decorative Arts              Pitcher     Covered Pitcher   
6406       American Decorative Arts                Plate               Plate   
7616       American Decorative Arts            Soup bowl           Soup Bowl   
15012      American Decorative Arts  Painting, miniature  Portrait of a Lady   

            Culture Artist Nationality  Object Begin Date  Object End Date  \
Object ID                                                                    
1674       American                NaN               1794           1799.0   
2430       American                NaN               1700           1900.0   
6406          Dutch                NaN               1740           1760.0   
7616       American                NaN               1885           1885.0   
15012      American                NaN               1840           1840.0   

                                            Medium  \
Object ID                                            
1674       Mahogany, ebony, ash, birch, white pine   
2430                                   Earthenware   
6406                                   Earthenware   
7616                                     Porcelain   
15012                          Watercolor on ivory   

                                         Credit Line        Country  
Object ID                                                            
1674         Friends of the American Wing Fund, 1962  United States  
2430                               Rogers Fund, 1913  United States  
6406                               Rogers Fund, 1933    Netherlands  
7616       Gift of Mr. and Mrs. Franklin Chace, 1969  United States  
15012                            Fletcher Fund, 2006            NaN

In [68]:
museum_df.shape

(204, 10)

* The Dataset has 204 total observations with total number of 11 columns(Including index).

# Step 2: Checking for duplicate observations. 

In [69]:
museum_df.index.duplicated(keep='first')

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

* Here, I have checked for duplicate observation. The return result with "True" as value are repeating observation.

In [70]:
#Found repetative unique identifier and observation values

museum_df[museum_df.index.duplicated(keep=False)]

Department Object Name  \
Object ID                                                          
100359                             Costume Institute         Hat   
100359                             Costume Institute         Hat   
316760     Arts of Africa, Oceania, and the Americas    Painting   
316760     Arts of Africa, Oceania, and the Americas    Painting   
486599                   Modern and Contemporary Art        Bowl   
486599                   Modern and Contemporary Art        Bowl   
563318                                  Egyptian Art   Jar Label   
563318                                  Egyptian Art   Jar Label   

                         Title        Culture Artist Nationality  \
Object ID                                                          
100359                     Hat         French      French|French   
100359                     Hat         French      French|French   
316760                Painting  Abelam people                NaN   
316760                Painting  Abelam people                NaN   
486599     "Tard la Nuit" Bowl            NaN            Italian   
486599     "Tard la Nuit" Bowl            NaN            Italian   
563318     Jar label fragments            NaN                NaN   
563318     Jar label fragments            NaN                NaN   

           Object Begin Date  Object End Date                  Medium  \
Object ID                                                               
100359                  1949           1949.0            wool, cotton   
100359                  1949           1949.0            wool, cotton   
316760                  1900           2000.0            Paper, paint   
316760                  1900           2000.0            Paper, paint   
486599                  1989           1989.0                   Glass   
486599                  1989           1989.0                   Glass   
563318                 -1390          -1390.0  Pottery and ink, paint   
563318                 -1390          -1390.0  Pottery and ink, paint   

                             Credit Line           Country  
Object ID                                                   
100359        Gift of Mary S. Ryan, 1981               NaN  
100359        Gift of Mary S. Ryan, 1981               NaN  
316760      Gift of Lita Osmundsen, 1992  Papua New Guinea  
316760      Gift of Lita Osmundsen, 1992  Papua New Guinea  
486599     Gift of Peter T. Joseph, 1995               NaN  
486599     Gift of Peter T. Joseph, 1995               NaN  
563318                 Rogers Fund, 1917             Egypt  
563318                 Rogers Fund, 1917             Egypt

* Checking for the duplicate observations by subsetting dataset.

In [71]:
#remving duplicate values by keeping the first occurance one and using inplace to reflect the changes into main memory

museum_df.drop_duplicates(subset=None, keep='first', inplace=True, ignore_index=False)

* Here the first occurance of the repeating observation is kept and set are dropped from the dataset to have **NO Redundant Data**.

In [72]:
#checking for any duplication after removing the repeating indexes observations.

museum_df.index.duplicated(keep='first')

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

* Rechecking for any more duplications. No more found.

museum_df.head()

In [73]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 199 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Department          198 non-null    object 
 1   Object Name         198 non-null    object 
 2   Title               189 non-null    object 
 3   Culture             81 non-null     object 
 4   Artist Nationality  85 non-null     object 
 5   Object Begin Date   199 non-null    int64  
 6   Object End Date     197 non-null    float64
 7   Medium              197 non-null    object 
 8   Credit Line         199 non-null    object 
 9   Country             30 non-null     object 
dtypes: float64(1), int64(1), object(8)
memory usage: 17.0+ KB


* Checking the datatypes of the features for getting brief information about the **Null data, and datatypes of the columns**. 

# Step 3: Object Begin and End date fields processing.

In [74]:
#checking converted datatype

museum_df.dtypes

Department             object
Object Name            object
Title                  object
Culture                object
Artist Nationality     object
Object Begin Date       int64
Object End Date       float64
Medium                 object
Credit Line            object
Country                object
dtype: object

* Checking the data type of the column **Object End Date** and **Object Begin Date**.

In [75]:
museum_df[museum_df['Object Begin Date'].isna() |
       museum_df['Object End Date'].isna()]

Department      Object Name  \
Object ID                                         
43092                Asian Art  Girdle ornament   
378518     Drawings and Prints          Drawing   

                                    Title Culture Artist Nationality  \
Object ID                                                              
43092                                 NaN   China                NaN   
378518     Design for Vis-à-vis, no. 3223     NaN           American   

           Object Begin Date  Object End Date  \
Object ID                                       
43092                   -206              NaN   
378518                  1876              NaN   

                                                      Medium  \
Object ID                                                      
43092                                                   Jade   
378518     Pen and black ink, watercolor and gouache with...   

                              Credit Line Country  
Object ID                                          
43092      Gift of Samuel T. Peters, 1913     NaN  
378518     Gift of William Brewster, 1923     NaN

* looking into the result of the info() method we come to know, there are some missing/NaN values in the column. Here I have filter the data to get subset having NaN values.

In [76]:
#checking for any observation with null values from columns Object Begin Date and Object End Date


#dropping such columns 
museum_df.drop(museum_df[museum_df['Object Begin Date'].isna() |
        museum_df['Object End Date'].isna()].index, axis=0,inplace=True)

* The filter applied is to get the subset of df that has null or missing value on either of one columns (Object begin date and Object end date).

* It's hard to determine informative data for missing values. Hence, I have dropped the 2 observations

**NOTE: I have not removed negative date year, B.C. year. as by observing  the relation of years with negative sign and "Obejct Date" column, it represent B. C. Year.**

In [77]:
#converting to "Object End Date" to int type 

museum_df['Object End Date'] = museum_df['Object End Date'].astype('int64')

**The datetime datatype has limitation in a range as it can not determine B.C. Years. Hence I have converted the datatype to int64 to make both date column have same type.**

In [78]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 197 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Department          196 non-null    object
 1   Object Name         196 non-null    object
 2   Title               188 non-null    object
 3   Culture             80 non-null     object
 4   Artist Nationality  84 non-null     object
 5   Object Begin Date   197 non-null    int64 
 6   Object End Date     197 non-null    int64 
 7   Medium              195 non-null    object
 8   Credit Line         197 non-null    object
 9   Country             30 non-null     object
dtypes: int64(2), object(8)
memory usage: 10.8+ KB


In [79]:
museum_df.drop(museum_df[museum_df['Object Begin Date']==0].index, axis=0,inplace=True)



In [80]:
museum_df.drop(museum_df[museum_df['Object End Date']==0].index, axis=0,inplace=True)

* It is been observed that there is 0 as a year in one of the observation, which make no contribution to the dataset. Hence, It is been dropped out of the dataset.

In [81]:
museum_df.head()

Department          Object Name               Title  \
Object ID                                                                      
1674       American Decorative Arts           Side Chair          Side Chair   
2430       American Decorative Arts              Pitcher     Covered Pitcher   
6406       American Decorative Arts                Plate               Plate   
7616       American Decorative Arts            Soup bowl           Soup Bowl   
15012      American Decorative Arts  Painting, miniature  Portrait of a Lady   

            Culture Artist Nationality  Object Begin Date  Object End Date  \
Object ID                                                                    
1674       American                NaN               1794             1799   
2430       American                NaN               1700             1900   
6406          Dutch                NaN               1740             1760   
7616       American                NaN               1885             1885   
15012      American                NaN               1840             1840   

                                            Medium  \
Object ID                                            
1674       Mahogany, ebony, ash, birch, white pine   
2430                                   Earthenware   
6406                                   Earthenware   
7616                                     Porcelain   
15012                          Watercolor on ivory   

                                         Credit Line        Country  
Object ID                                                            
1674         Friends of the American Wing Fund, 1962  United States  
2430                               Rogers Fund, 1913  United States  
6406                               Rogers Fund, 1933    Netherlands  
7616       Gift of Mr. and Mrs. Franklin Chace, 1969  United States  
15012                            Fletcher Fund, 2006            NaN

# Step 4:Processing Department Feature

In [82]:
#checking for null values and datatype of Department column

museum_df.info()

#It has 1 null value observation and need to type cast to Categorical

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Department          195 non-null    object
 1   Object Name         195 non-null    object
 2   Title               187 non-null    object
 3   Culture             80 non-null     object
 4   Artist Nationality  84 non-null     object
 5   Object Begin Date   196 non-null    int64 
 6   Object End Date     196 non-null    int64 
 7   Medium              194 non-null    object
 8   Credit Line         196 non-null    object
 9   Country             29 non-null     object
dtypes: int64(2), object(8)
memory usage: 13.8+ KB


* Checking brief information about the column Department. It is type Object and has 1 missing value.

In [83]:
#checking the NaN valued observation

museum_df[museum_df['Department'].isnull()]

Department Object Name           Title  Culture Artist Nationality  \
Object ID                                                                      
317005           NaN       Cloth  Carrying Cloth  Quechua                NaN   

           Object Begin Date  Object End Date        Medium  \
Object ID                                                     
317005                  1900             2000  Camelid hair   

                                    Credit Line Country  
Object ID                                                
317005     Bequest of Arthur M. Bullowa, 1993      Peru

* **The result shows the observation have no Department value.**

In [84]:
#checking Department types

museum_df['Department'].value_counts()

Drawings and Prints                          78
Costume Institute                            19
Photographs                                  15
European Sculpture and Decorative Arts       14
Greek and Roman Art                          11
Asian Art                                    10
Egyptian Art                                 10
Arms and Armor                                7
Arts of Africa, Oceania, and the Americas     6
Modern and Contemporary Art                   6
American Decorative Arts                      5
Islamic Art                                   4
Ancient Near Eastern Art                      3
Medieval Art                                  3
The Cloisters                                 3
Robert Lehman Collection                      1
Name: Department, dtype: int64

In [85]:
museum_df['Department'].fillna('Others',inplace=True)

* Filling NaN with value **"Others"** by adding new department

In [86]:
museum_df['Department'] = museum_df['Department'].astype('category')

* converting datatype from **Object to categorical**

In [87]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         195 non-null    object  
 2   Title               187 non-null    object  
 3   Culture             80 non-null     object  
 4   Artist Nationality  84 non-null     object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              194 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             29 non-null     object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


* verifying the updated changes.

In [88]:
museum_df['Department']

Object ID
1674      American Decorative Arts
2430      American Decorative Arts
6406      American Decorative Arts
7616      American Decorative Arts
15012     American Decorative Arts
                    ...           
779306         Drawings and Prints
785013         Drawings and Prints
813925         Drawings and Prints
820618         Drawings and Prints
824428         Drawings and Prints
Name: Department, Length: 196, dtype: category
Categories (17, object): ['American Decorative Arts', 'Ancient Near Eastern Art', 'Arms and Armor', 'Arts of Africa, Oceania, and the Americas', ..., 'Others', 'Photographs', 'Robert Lehman Collection', 'The Cloisters']

#  Step 5:Processing Object Name Feature 

In [89]:
museum_df['Object Name'] = museum_df['Object Name'].str.split(',').str[0].str.replace(';',' ').str.strip()

* replaceing ';' with whitespace, spliting string before ',' and removing whitespace from end or begning 

In [90]:
museum_df.head(100)

Department                      Object Name  \
Object ID                                                              
1674       American Decorative Arts                       Side Chair   
2430       American Decorative Arts                          Pitcher   
6406       American Decorative Arts                            Plate   
7616       American Decorative Arts                        Soup bowl   
15012      American Decorative Arts                         Painting   
...                             ...                              ...   
365646          Drawings and Prints                          Drawing   
367617          Drawings and Prints  Drawing Ornament & Architecture   
371229          Drawings and Prints                            Print   
377953          Drawings and Prints                             Book   
378914          Drawings and Prints                          Drawing   

                                                       Title   Culture  \
Object ID                                                                
1674                                              Side Chair  American   
2430                                         Covered Pitcher  American   
6406                                                   Plate     Dutch   
7616                                               Soup Bowl  American   
15012                                     Portrait of a Lady  American   
...                                                      ...       ...   
365646                                            Saint Paul       NaN   
367617               Francis W. Little House: Furniture Plan       NaN   
371229     Theodoric IV from the series Counts and Counte...       NaN   
377953                L'Art de bâtir des maisons de Campagne       NaN   
378914                     Design for Pony Phaeton, no. 3162       NaN   

                           Artist Nationality  Object Begin Date  \
Object ID                                                          
1674                                      NaN               1794   
2430                                      NaN               1700   
6406                                      NaN               1740   
7616                                      NaN               1885   
15012                                     NaN               1840   
...                                       ...                ...   
365646                                Italian               1591   
367617                               American               1908   
371229                      Dutch|Dutch|Dutch               1650   
377953     French|French|French|French|French               1743   
378914                               American               1875   

           Object End Date                                             Medium  \
Object ID                                                                       
1674                  1799            Mahogany, ebony, ash, birch, white pine   
2430                  1900                                        Earthenware   
6406                  1760                                        Earthenware   
7616                  1885                                          Porcelain   
15012                 1840                                Watercolor on ivory   
...                    ...                                                ...   
365646                1666  Pen and brown ink, brush and brown ink, over l...   
367617                1914               Graphite and orange and brown pencil   
371229                1650       Engraving and etching; second state of three   
377953                1743           Printed book with engraved illustrations   
378914                1875          Pen and black ink, watercolor and gouache   

                                             Credit Line        Country  
Object ID                                                                
1674             Friends of the Americ

In [91]:
museum_df[museum_df['Object Name'].isna()]

Department Object Name  \
Object ID                                            
656878     Modern and Contemporary Art         NaN   

                                                       Title Culture  \
Object ID                                                              
656878     New York City 2[unfinished, formerly New York ...     NaN   

          Artist Nationality  Object Begin Date  Object End Date  \
Object ID                                                          
656878                   NaN               1941             1941   

                           Medium  \
Object ID                           
656878     Oil and tape on canvas   

                                                 Credit Line Country  
Object ID                                                             
656878     San Francisco Museum of Modern Art, Purchase t...     USA

In [92]:
museum_df['Object Name'].fillna("Not Available",inplace = True)

* Filling NaN value with **"Not Available value"** to handle missing data

In [93]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               187 non-null    object  
 3   Culture             80 non-null     object  
 4   Artist Nationality  84 non-null     object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              194 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             29 non-null     object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


* verifying the chnages: Non-null count  and data type.

# Step 6: Processing Title Feature

* In the below code, I have checkig for null/NaN values in Title column.

In [94]:
museum_df[museum_df['Title'].isna()]

Department                           Object Name Title  \
Object ID                                                          
39917      Asian Art  Hanging scroll mounted as handscroll   NaN   
53280      Asian Art                              Bracelet   NaN   
54547      Asian Art                                 Print   NaN   
54626      Asian Art                            Album leaf   NaN   
55442      Asian Art                                 Print   NaN   
58899      Asian Art                                  Inrō   NaN   
62174      Asian Art                                Basket   NaN   
67976      Asian Art                                 Piece   NaN   
77166      Asian Art                               Drawing   NaN   

                               Culture Artist Nationality  Object Begin Date  \
Object ID                                                                      
39917                            China            Chinese               1037   
53280                         Thailand                NaN               -300   
54547                            Japan           Japanese               1800   
54626                            Japan           Japanese               1877   
55442                            Japan           Japanese               1838   
58899                            Japan                NaN               1800   
62174                            Japan                NaN               1800   
67976                            Japan                NaN               1800   
77166      India (Guler, Punjab Hills)                NaN               1790   

           Object End Date                                             Medium  \
Object ID                                                                       
39917                 1101  Hanging scroll mounted as a handscroll; ink on...   
53280                  400                                             Bronze   
54547                 1899  Polychrome woodblock print (surimono); ink and...   
54626                 1897                  Album leaf; ink and color on silk   
55442                 1838  Polychrome woodblock print; ink and color on p...   
58899                 1899  Lacquer, hirame ground, gold, red, silver and ...   
62174                 1899                                             Bamboo   
67976                 1899                                    Silk / Brocaded   
77166                 1810                                       Ink on paper   

                                                 Credit Line Country  
Object ID                                                             
39917                  Bequest of John M. Crawford Jr., 1988     NaN  
53280      Samuel Eilenberg Collection, Bequest of Samuel...     NaN  
54547      H. O. Havemeyer Collection, Bequest of Mrs. H....     NaN  
54626      Charles Stewart Smith Collection, Gift of Mrs....     NaN  
55442                                      Rogers Fund, 1922     NaN  
58899      The Howard Mansfield Collection, Purchase, Rog...     NaN  
62174      Edward C. Moore Collection, Bequest of Edward ...     NaN  
67976                                      Rogers Fund, 1909     NaN  
77166                     Gift of Cynthia Hazen Polsky, 2011     NaN

* **Subset of Nan valued dataset for column Object Name.**

In [95]:
#Adding object name as title 

museum_df['Title']= np.where(museum_df['Title'].isnull(), museum_df['Object Name'], museum_df['Title'])

* The null values are replpaceed with the Title value of the resective column as by observing the dataset, It can be said the Object name and Title are similar. 

In [96]:
#no null data left 

museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               196 non-null    object  
 3   Culture             80 non-null     object  
 4   Artist Nationality  84 non-null     object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              194 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             29 non-null     object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


In [97]:
#checking for strings with "[]" in a column 

museum_df[museum_df['Title'].str.startswith("[")]

Department                  Object Name  \
Object ID                                             
262991     Photographs                   Photograph   
265758     Photographs                   Photograph   
266167     Photographs                   Photograph   
273592     Photographs                     Negative   
276937     Photographs                     Negative   
670640     Photographs  Photograph  Photomicrograph   
679516     Photographs                        Album   
732391     Photographs                     Ephemera   

                                                       Title Culture  \
Object ID                                                              
262991     [Detail of Vanitas Figures Carved on the Choir...     NaN   
265758     [Graffiti on Backstage Door of the School of A...     NaN   
266167                          [Study of Australian Grapes]     NaN   
273592                                         [Peggy Bacon]     NaN   
276937     [Two 35mm Film Frames: Boys Playing on Stoop N...     NaN   
670640                                        [Snow Crystal]     NaN   
679516                                               [Album]     NaN   
732391                 [Rusted Metal Sign: "Slow, Children"]     NaN   

          Artist Nationality  Object Begin Date  Object End Date  \
Object ID                                                          
262991                   NaN               1860             1869   
265758              American               1945             1945   
266167                   NaN               1858             1862   
273592              American               1934             1934   
276937              American               1933             1934   
670640              American               1890             1929   
679516                   NaN               1860             1879   
732391                   NaN               1960             1975   

                                             Medium  \
Object ID                                             
262991     Albumen silver print from glass negative   
265758                         Gelatin silver print   
266167     Albumen silver print from glass negative   
273592                                Film negative   
276937                                Film negative   
670640                         Gelatin silver print   
679516                                          NaN   
732391                                Painted metal   

                                                 Credit Line Country  
Object ID                                                             
262991                      Gift of Lucien Goldschmidt, 1981     NaN  
265758     Gift of Paul F. Walter, in memory of Christoph...     NaN  
266167     Purchase, The Horace W. Goldsmith Foundation G...     NaN  
273592                            Walker Evans Archive, 1994     NaN  
276937                            Walker Evans Archive, 1994     NaN  
670640        Purchase, Alfred Stieglitz Society Gifts, 2015     NaN  
679516                             Gift of Eva Kasiska, 1947     NaN  
732391                            Walker Evans Archive, 1994     NaN

* The column **Title** has some values enclosed in "[]". Hence,  the result shows the removal of brackets to standardise the data.

In [98]:
museum_df['Title'] =  museum_df['Title'].apply(lambda x: x.replace('[','').replace(']','')) 

In [99]:
#checking for string with "[]" after removing.

museum_df[museum_df['Title'].str.startswith("[")]

Empty DataFrame
Columns: [Department, Object Name, Title, Culture, Artist Nationality, Object Begin Date, Object End Date, Medium, Credit Line, Country]
Index: []

* Verifying the changes by checking no values have "[]"

In [100]:
#removing whitespace from the Title column values

museum_df['Title'].str.strip()

Object ID
1674                                             Side Chair
2430                                        Covered Pitcher
6406                                                  Plate
7616                                              Soup Bowl
15012                                    Portrait of a Lady
                                ...                        
779306    Packard-Tourist, from the "Antique Autos" seri...
785013    Norwegian Air Force In Canada Patrol Bomber (N...
813925                                           Stonehenge
820618    Psyche, holding the flask of beauty being tran...
824428                                 Landscape and Cattle
Name: Title, Length: 196, dtype: object

# Step 7: Processing Culture Column

In [101]:
museum_df['Culture']

Object ID
1674      American
2430      American
6406         Dutch
7616      American
15012     American
            ...   
779306         NaN
785013         NaN
813925         NaN
820618         NaN
824428         NaN
Name: Culture, Length: 196, dtype: object

In [102]:
#adding "Other" in Culture NaN valued observatios

museum_df['Culture'].fillna("Others",inplace=True)

* Handaling missing/NaN valued data by filling "Others". It is not advisable to drop for avoiding over fitting of data problem.

In [103]:
#removing string after first occurance of ',', removing whitespace and also removing possibly and probably words to make consistent

museum_df['Culture'] = museum_df['Culture'].str.split(',').str[0].str.replace('possibly','').str.replace('probably','').str.strip()

In [104]:
#converting to type category

museum_df['Culture'] = museum_df['Culture'].astype("category")

In [105]:
#replacing "Japanese" to "Japan" from list of culture

museum_df['Culture'] = museum_df['Culture'].str.replace('Japanese','Japan')

museum_df['Culture'] = museum_df['Culture'].str.replace('Italic','Italian')

* Removing same names with different types to optimize the list of country and make it consistent

In [106]:
#reformating "India" as a value

museum_df['Culture'] = museum_df['Culture'].str.split(' \(').str[0]
museum_df['Culture']

Object ID
1674      American
2430      American
6406         Dutch
7616      American
15012     American
            ...   
779306      Others
785013      Others
813925      Others
820618      Others
824428      Others
Name: Culture, Length: 196, dtype: object

# Step 8: Processing Artist  Nationality column

In [107]:
museum_df['Artist Nationality']

Object ID
1674      NaN
2430      NaN
6406      NaN
7616      NaN
15012     NaN
         ... 
779306    NaN
785013    NaN
813925    NaN
820618    NaN
824428    NaN
Name: Artist Nationality, Length: 196, dtype: object

In [108]:
#filling NaN with 'unknown'

museum_df['Artist Nationality'].fillna("Unknown",inplace=True)

* The missing values in **Artist Nationality** has been handled by filling "Unknown" as a value.

In [109]:
#Converting to category

museum_df['Artist Nationality'] = museum_df['Artist Nationality'].astype("category")

In [110]:
museum_df['Artist Nationality']

Object ID
1674      Unknown
2430      Unknown
6406      Unknown
7616      Unknown
15012     Unknown
           ...   
779306    Unknown
785013    Unknown
813925    Unknown
820618    Unknown
824428    Unknown
Name: Artist Nationality, Length: 196, dtype: category
Categories (37, object): ['American', 'American, born Russia', 'American|American', 'American|American|American', ..., 'Mexican', 'Netherlandish|German', 'Scottish|British|British, Scottish|British|Br..., 'Unknown']

In [111]:
#fetching string before special character '|' or ','


museum_df['Artist Nationality'] = museum_df['Artist Nationality'].str.split('|').str[0].str.split(',').str[0]
museum_df['Artist Nationality']

Object ID
1674      Unknown
2430      Unknown
6406      Unknown
7616      Unknown
15012     Unknown
           ...   
779306    Unknown
785013    Unknown
813925    Unknown
820618    Unknown
824428    Unknown
Name: Artist Nationality, Length: 196, dtype: object

* Updating values such **French | French** and **America|America|America**.

# Step 9: Processing Medium Column

In [112]:
museum_df['Medium']

Object ID
1674      Mahogany, ebony, ash, birch, white pine
2430                                  Earthenware
6406                                  Earthenware
7616                                    Porcelain
15012                         Watercolor on ivory
                           ...                   
779306                Commercial color lithograph
785013                Commercial color lithograph
813925          Sugar-lift aquatint with roulette
820618                                  Engraving
824428                                    Etching
Name: Medium, Length: 196, dtype: object

In [113]:
#checking for nan values

museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               196 non-null    object  
 3   Culture             196 non-null    object  
 4   Artist Nationality  196 non-null    object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              194 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             29 non-null     object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


In [114]:
# removing square brackets from string

museum_df['Medium'] = museum_df['Medium'].str.replace('[','').str.replace(']','')


C:\Users\admin\AppData\Local\Temp/ipykernel_19188/983856892.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  museum_df['Medium'] = museum_df['Medium'].str.replace('[','').str.replace(']','')


In [115]:
#checking for any value starting with '[' and ending with']'

museum_df['Medium'].str.startswith("[")

Object ID
1674      False
2430      False
6406      False
7616      False
15012     False
          ...  
779306    False
785013    False
813925    False
820618    False
824428    False
Name: Medium, Length: 196, dtype: object

In [116]:
museum_df['Medium'].fillna('no medium available',inplace=True)

* filling null values with No Medium Available. 

In [117]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               196 non-null    object  
 3   Culture             196 non-null    object  
 4   Artist Nationality  196 non-null    object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              196 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             29 non-null     object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


In [118]:
museum_df['Medium']

Object ID
1674      Mahogany, ebony, ash, birch, white pine
2430                                  Earthenware
6406                                  Earthenware
7616                                    Porcelain
15012                         Watercolor on ivory
                           ...                   
779306                Commercial color lithograph
785013                Commercial color lithograph
813925          Sugar-lift aquatint with roulette
820618                                  Engraving
824428                                    Etching
Name: Medium, Length: 196, dtype: object

In [119]:
#museum_df['Medium'] = museum_df.assign(museum_df=museum_df['Medium'].str.split('; ')).explode('Medium')

In [120]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               196 non-null    object  
 3   Culture             196 non-null    object  
 4   Artist Nationality  196 non-null    object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              196 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             29 non-null     object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


# Step 10: Processing Country Column

In [121]:
museum_df['Country']

Object ID
1674      United States
2430      United States
6406        Netherlands
7616      United States
15012               NaN
              ...      
779306              NaN
785013              NaN
813925              NaN
820618              NaN
824428              NaN
Name: Country, Length: 196, dtype: object

In [122]:
#filling missing values with "Unknown" as value

museum_df['Country'].fillna('Unknown',inplace= True)

* The Country column has 60-70% of missing values. To handle null value, "Unknown" is filled in as value. 

In [123]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               196 non-null    object  
 3   Culture             196 non-null    object  
 4   Artist Nationality  196 non-null    object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              196 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             196 non-null    object  
dtypes: category(1), int64(2), object(7)
memory usage: 13.6+ KB


In [124]:
#replacing "present-day " with '' for standardizing it and remove whitespace from start and end.

museum_df['Country'] = museum_df['Country'].str.replace('present-day ','').str.strip()

In [125]:
museum_df['Country'] = museum_df['Country'].str.replace('United States','USA')

In [126]:
museum_df['Country'] = museum_df['Country'].astype("category")

* Removing additional words such as: **Present-day, United States, and Probably**

In [127]:
museum_df['Country']

Object ID
1674              USA
2430              USA
6406      Netherlands
7616              USA
15012         Unknown
             ...     
779306        Unknown
785013        Unknown
813925        Unknown
820618        Unknown
824428        Unknown
Name: Country, Length: 196, dtype: category
Categories (13, object): ['Costa Rica', 'Egypt', 'France', 'India', ..., 'Peru', 'USA', 'Unknown', 'Uzbekistan']

# Exporting Cleaned Dataset

**After cleaning the dataset by undergoing all the above mentioned steps. Here, I have verified the Null counts and the data types.**

In [128]:
museum_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 196 entries, 1674 to 824428
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Department          196 non-null    category
 1   Object Name         196 non-null    object  
 2   Title               196 non-null    object  
 3   Culture             196 non-null    object  
 4   Artist Nationality  196 non-null    object  
 5   Object Begin Date   196 non-null    int64   
 6   Object End Date     196 non-null    int64   
 7   Medium              196 non-null    object  
 8   Credit Line         196 non-null    object  
 9   Country             196 non-null    category
dtypes: category(2), int64(2), object(6)
memory usage: 13.3+ KB


**Here I have exported the cleaned/ processed dataset**

In [129]:
museum_df.to_csv('cleaned_assignment_8717454.csv') # Writes to a CSV file

* **Note:** The file will be created and stored in the same directory of this file.